packages/cli/src/controllers/auth.controller.ts

```diff
@@ -233,4 +254,25 @@ export class AuthController {
 		res.clearCookie(AUTH_COOKIE_NAME);
 		return { loggedOut: true };
 	}
+
+	private async validateMfaToken(user: User, token: string) {
+		return this.mfaService.totp.verifySecret({
+			secret: user.mfaSecret ?? '',
+			token,
+		});
+	}
+
+	private async validateMfaRecoveryCode(user: User, mfaRecoveryCode: string) {
+		const index = user.mfaRecoveryCodes.indexOf(mfaRecoveryCode);
+		if (index === -1) return false;
+
+		// remove used recovery code
+		user.mfaRecoveryCodes.splice(index, 1);
```

Comment:
nitpick - always prefer non-destructive functional code...
```suggestion
	         if (!user.mfaRecoveryCodes.indexOf(mfaRecoveryCode)) {
	            return false;
	         }
                 const remaining = user.mfaRecoveryCodes.filter((code) => code !== mfaRecoveryCode);
```